### ISSUE TITLE
Function Hook to CInputManager::getMouseCoordsInternal only works in debug mode.

### ISSUE THREAD

**Issue by raybbian** [hyprwm/Hyprland#8845]
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

No

### System Info and Version

[tmp.txt](https://github.com/user-attachments/files/18243977/tmp.txt)

### Description

I'm developing a plugin that wants to hook into `getMouseCoordsInternal` to change the position of the mouse when some conditions are met. 

On a nested debug build of Hyprland, my plugin is able to hook just fine. However, on nested release builds of Hyprland, the API reports a successful hook, but the hook is not called. On non-nested release Hyprland, any attempts to call the original function through `m_pOriginal` of `CFunctionHook` causes a segfault in the plugin, which unloads the plugin.

Here is how I am hooking the functions:
```cpp
static void initFunctions() {
    static auto FNS =
        HyprlandAPI::findFunctionsByName(PHANDLE, "renderWorkspace");
    if (FNS.empty()) {
        failNotification("No fns for hook renderWorkspace!");
        throw std::runtime_error(
            "[Hyprtasking] No fns for hook renderWorkspace");
    }
    g_pRenderWorkspaceHook = HyprlandAPI::createFunctionHook(
        PHANDLE, FNS[0].address, (void *)hkRenderWorkspace);

    FNS = HyprlandAPI::findFunctionsByName(PHANDLE, "getMouseCoordsInternal");
    if (FNS.empty()) {
        failNotification("No fns for hook getMouseCoordsInternal");
        throw std::runtime_error(
            "[Hyprtasking] No fns for hook getMouseCoordsInternal");
    }
    g_pGetMouseCoordsInternalHook = HyprlandAPI::createFunctionHook(
        PHANDLE, FNS[0].address, (void *)hkGetMouseCoordsInternal);

    bool success = g_pRenderWorkspaceHook->hook();
    success = success && g_pGetMouseCoordsInternalHook->hook();
    if (!success) {
        failNotification("Failed initializing hooks");
        throw std::runtime_error("[Hyprtasking] Failed initializing hooks");
    }
}
```

The function that should execute: (No logs are printed in release builds, while they are printed in Debug builds)
```cpp
static Vector2D hkGetMouseCoordsInternal(void *thisptr) {
    Debug::log(LOG, std::format("[Hyprtasking] getMouseCoords hook called"));
    const Vector2D oMousePos =
        ((tGetMouseCoordsInternal)(g_pGetMouseCoordsInternalHook->m_pOriginal))(
            thisptr);
    const PHLMONITOR pMonitor = g_pCompositor->getMonitorFromVector(oMousePos);
    const auto view = getViewForMonitor(pMonitor);

    if (pMonitor == nullptr || view == nullptr || !view->isActive())
        return oMousePos;

    const Vector2D newPos = view->mouseCoordsWorkspaceRelative(oMousePos);
    Debug::log(
        LOG, std::format(
                 "[Hyprtasking] mouse coords hooked from ({}, {}) to ({}, {})",
                 oMousePos.x, oMousePos.y, newPos.x, newPos.y));
    return newPos;
}
```

What causes the segfault in non-nested release: 
```cpp
    const Vector2D mousePos =
        ((tGetMouseCoordsInternal)(g_pGetMouseCoordsInternalHook->m_pOriginal))(
            g_pInputManager.get());
```

### How to reproduce

1. Create a plugin that hooks `getMouseCoordsInternal` (Can attempt with https://github.com/raybbian/hyprtasking)
2. Load the plugin on release build Hyprland
3. Call the original function (will happen on mouse move after dispatching `hyprtasking:toggle`)
    a. Also note that hook is not called
4. Segfault

### Crash reports, logs, images, videos

Nested release:
https://github.com/user-attachments/assets/0faed1ff-46de-41ee-85e3-287863feafef

Nested debug:
https://github.com/user-attachments/assets/33789962-d09d-4295-b9d5-68a7c0391bf9

Release:
https://github.com/user-attachments/assets/7fb4330b-dea3-4147-9762-57b8f822dba4

**Comment by raybbian** [comment]
Function getting inlined maybe?

**Comment by vaxerski** [comment]
can you install the hook on a release version, attach gdb to the running hyprland (you can do it via ssh, `sudo gdb attach $(pidof Hyprland)` and run `disassemble _ZN13CInputManager22getMouseCoordsInternalEv`?

**Comment by raybbian** [comment]
yeah, one sec

**Comment by vaxerski** [comment]
possible we tripped CFI. Try building hyprland with `-fcf-protection=none` and see if it gets rid of the endbr64

**Comment by raybbian** [comment]
I should clarify: my "non-nested release" Hyprland is installed from pacman:
```
[rayb@rayb-linux build]$ which Hyprland
/usr/bin/Hyprland
[rayb@rayb-linux build]$ hyprland -v
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.3.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.1


flags set:
debug
```

Nested release hyprland is build from hyprland-git with the appropriate commit checked out:
```
[rayb@rayb-linux build]$ which ./Hyprland 
/home/rayb/Projects/hyprland-git/build/Hyprland
[rayb@rayb-linux build]$ ./Hyprland -v
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887 dirty (version: bump to 0.46.2).
Date: Thu Dec 19 14:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.2.6
 hyprcursor 0.1.10
 hyprgraphics 0.1.1


flags set:
debug
```

Compiling hyprland from the repository with `-fcf-protection=none` yields the same asm as previous:
```

**Comment by vaxerski** [comment]
I assume without the endbr64 at the beginning, the hook works?

**Comment by raybbian** [comment]
No - on "nested release" my hook is never called as if the function wasn't hooked in the first place

**Comment by vaxerski** [comment]
Can you show the disassembly of nested release (where it doesnt work) where you already _loaded and applied your hook and it doesnt work_?

**Comment by raybbian** [comment]
```
(gdb) disassemble _ZN13CInputManager22getMouseCoordsInternalEv
Dump of assembler code for function _ZN13CInputManager22getMouseCoordsInternalEv:
   0x000058d991b07dc0 <+0>:	movabs $0x74a7582c7f8e,%rax
   0x000058d991b07dca <+10>:	jmp    *%rax
   0x000058d991b07dcc <+12>:	pop    %rax
   0x000058d991b07dcd <+13>:	nop
   0x000058d991b07dce <+14>:	nop
   0x000058d991b07dcf <+15>:	nop
   0x000058d991b07dd0 <+16>:	nop
   0x000058d991b07dd1 <+17>:	nop
   0x000058d991b07dd2 <+18>:	nop
   0x000058d991b07dd3 <+19>:	nop
   0x000058d991b07dd4 <+20>:	nop
   0x000058d991b07dd5 <+21>:	mov    %rbx,0x8(%rsp)
   0x000058d991b07dda <+26>:	mov    %rdi,%rbx
   0x000058d991b07ddd <+29>:	call   0x58d991acbd30 <_ZN15CPointerManager8positionEv>
   0x000058d991b07de2 <+34>:	mov    0x8(%rsp),%rax
   0x000058d991b07de7 <+39>:	sub    %fs:0x28,%rax
   0x000058d991b07df0 <+48>:	jne    0x58d991b07dfb <_ZN13CInputManager22getMouseCoordsInternalEv+59>
   0x000058d991b07df2 <+50>:	add    $0x10,%rsp
   0x000058d991b07df6 <+54>:	mov    %rbx,%rax
   0x000058d991b07df9 <+57>:	pop    %rbx
   0x000058d991b07dfa <+58>:	ret
   0x000058d991b07dfb <+59>:	call   0x58d991850980 <__stack_chk_fail@plt>
End of assembler dump.
```

**Comment by vaxerski** [comment]
address looks very bad. Can you post a log? Also, what does `cat /proc/$(pidof Hyprland)/maps` return?

**Comment by raybbian** [comment]
Log: 
[log.txt](https://github.com/user-attachments/files/18247049/log.txt)
Maps: 
[maps.txt](https://github.com/user-attachments/files/18247050/maps.txt)

*should* be same process

**Comment by vaxerski** [comment]
```
[LOG] getAddressForTrampo: Returning addr 0x5826dd20c000 for page at 0x5826dd20c000
[LOG] getAddressForTrampo: Returning addr 0x5826dd20c040 for page at 0x5826dd20c000
```

I don't think it's the same...? You can see the addr is `0x7038940999a6` in the disasm...

**Comment by raybbian** [comment]
Let me try again...

Log:
TODO
Maps:
TODO
Disasm:
```
TODO
```

**Comment by vaxerski** [comment]
ah right shit sorry my bad, was reading wrong, xmas moment. Can you please follow the trampoline and disassemble 40 bytes of it? (You can find it in the log, in `[LOG] getAddressForTrampo: Returning addr 0x63ea08f2b000 for page at 0x63ea08f2b000` the trampo is the "returning addr ...")

I am pretty sure the trampoline has to be bad if calling original crashes.

Also, can you try _not_ calling the original and seeing if that works? Just return some dummy 1,1 for testing.

**Comment by raybbian** [comment]
> I am pretty sure the trampoline has to be bad if calling original crashes.

The original crashes only on non-nested Hyprland, meaning it could have also been `endbr64` (?)

> Also, can you try not calling the original and seeing if that works? Just return some dummy 1,1 for testing.

Here is the hook
```cpp
static Vector2D hkGetMouseCoordsInternal(void *thisptr) {
    Debug::log(LOG, "[Hyprtasking] Hooked get mouse coords internal");
    return {1, 1};
}
```
and trampoline `0x5b60e4931040` from 
```
[LOG] getAddressForTrampo: Returning addr 0x5b60e4931040 for page at 0x5b60e4931000
[LOG] [Hyprtasking] Hooked _ZN13CInputManager22getMouseCoordsInternalEv
```
with disasm
```
(gdb) x/16i 0x5b60e4931040
   0x5b60e4931040:	push   %rbx
   0x5b60e4931041:	sub    $0x10,%rsp
   0x5b60e4931045:	
    mov    -0xd894(%rip),%rsi        # 0x5b60e49237b8 <g_pPointerManager>
   0x5b60e493104c:	mov    %fs:0x28,%rbx
   0x5b60e4931055:	push   %rax
   0x5b60e4931056:	movabs $0x5b60e446adcc,%rax
   0x5b60e4931060:	jmp    *%rax
   0x5b60e4931062:	add    %al,(%rax)
   0x5b60e4931064:	add    %al,(%rax)
   0x5b60e4931066:	add    %al,(%rax)
   0x5b60e4931068:	add    %al,(%rax)
   0x5b60e493106a:	add    %al,(%rax)
   0x5b60e493106c:	add    %al,(%rax)
   0x5b60e493106e:	add    %al,(%rax)
   0x5b60e4931070:	add    %al,(%rax)
   0x5b60e4931072:	add    %al,(%rax)
```

... I've also just realized that the hook does get called (sometimes). Whoops.

The hook gets triggered when I left click on my plugin's overlay. This is the code that executes when that happens:
```cpp
static void onMouseButton(void *thisptr, SCallbackInfo &info, std::any args) {
    const PHLMONITOR pMonitor = g_pCompositor->getMonitorFromCursor();
    const auto view = getViewForMonitor(pMonitor);
    if (pMonitor == nullptr || view == nullptr || !view->isActive())
        return;

    info.cancelled = true;

    const auto e = std::any_cast<IPointer::SButtonEvent>(args);
    if (e.button != BTN_LEFT)
        return;
    const bool pressed = e.state == WL_POINTER_BUTTON_STATE_PRESSED;
    if (pressed) {
        g_pKeybindManager->changeMouseBindMode(MBIND_MOVE);
    } else {
        g_pKeybindManager->changeMouseBindMode(MBIND_INVALID);
    }
}
```
`changeMouseBindMode` *does* call `getMouseCoordsInternal`, so I believe the hook itself works fine (?)

However, the hook is *not called* whenever the mouse moves. This is odd because I have a `mouseMove` callback `onMouseMove` that *is* being called from (I believe) [here.](https://github.com/hyprwm/Hyprland/blob/dddb64dc353ec9ef73be22d1cc1441782fb44068/src/managers/input/InputManager.cpp#L179)

To call `onMouseMove`, doesn't  `mouseMoveUnified` need to call `getMouseCoordsInternal()` first?

```cpp
void CInputManager::mouseMoveUnified(uint32_t time, bool refocus, bool mouse) {
    m_bLastInputMouse = mouse;

    if (!g_pCompositor->m_bReadyToProcess || g_pCompositor->m_bIsShuttingDown || g_pCompositor->m_bUnsafeState)
        return;

    Vector2D const mouseCoords        = getMouseCoordsInternal();
    auto const     MOUSECOORDSFLOORED = mouseCoords.floor();

    // ...   

    EMIT_HOOK_EVENT_CANCELLABLE("mouseMove", MOUSECOORDSFLOORED);
}
```

... what?

This is likely a skill issue on my part - thanks for your help

happy holidays!

**Comment by vaxerski** [comment]
It genuinely isn't - this is unfortunately the tradeoff of hooks. Assembly debugging is very non-trivial :P

Anyways, I assume that it _is_ being called in debug on mouse move.

> The original crashes only on non-nested Hyprland, meaning it could have also been endbr64 (?)

Let's get this straight:
Release, non-nested:
 - original crashes
 - hook doesnt get called on mouse move

Debug, nested:
 - original works
 - hook gets called on mouse move

Correct?

Now, is "release non nested" a "git, with fcf-protection=none", or "pulled from arch repos"?

If it's arch repos, try release, non-nested, with fcf-protection=none.

**Comment by raybbian** [comment]
> Anyways, I assume that it is being called in debug on mouse move.

Yep

> Correct?

That is the case. "Original crashes" is a SIGSEGV in `onMouseMove` (because my `onMouseMove` calls the original `GetMouseCoordsInternal`), which causes Hyprland to unload the plugin.

> Now, is "release non nested" a "git, with fcf-protection=none", or "pulled from arch repos"?

"Release non-nested" is the hyprland arch package.

> If it's arch repos, try release, non-nested, with fcf-protection=none.

TODO

**Comment by vaxerski** [comment]
I have a feeling it might actually be inlining, but that wouldn't explain the crashes.

**Comment by raybbian** [comment]
Disasm before plugin load
```
(gdb) disassemble _ZN13CInputManager22getMouseCoordsInternalEv
Dump of assembler code for function _ZN13CInputManager22getMouseCoordsInternalEv:
   0x000059e3a2b3edc0 <+0>:     push   %rbx
   0x000059e3a2b3edc1 <+1>:     sub    $0x10,%rsp
   0x000059e3a2b3edc5 <+5>:     mov    0x4b89ec(%rip),%rsi        # 0x59e3a2ff77b8 <g_pPointerManager>
   0x000059e3a2b3edcc <+12>:    mov    %fs:0x28,%rbx
   0x000059e3a2b3edd5 <+21>:    mov    %rbx,0x8(%rsp)
   0x000059e3a2b3edda <+26>:    mov    %rdi,%rbx
   0x000059e3a2b3eddd <+29>:    call   0x59e3a2b02d30 <_ZN15CPointerManager8positionEv>
   0x000059e3a2b3ede2 <+34>:    mov    0x8(%rsp),%rax
   0x000059e3a2b3ede7 <+39>:    sub    %fs:0x28,%rax
   0x000059e3a2b3edf0 <+48>:    jne    0x59e3a2b3edfb <_ZN13CInputManager22getMouseCoordsInternalEv+59>
   0x000059e3a2b3edf2 <+50>:    add    $0x10,%rsp
   0x000059e3a2b3edf6 <+54>:    mov    %rbx,%rax
   0x000059e3a2b3edf9 <+57>:    pop    %rbx
   0x000059e3a2b3edfa <+58>:    ret
   0x000059e3a2b3edfb <+59>:    call   0x59e3a2887980 <__stack_chk_fail@plt>
End of assembler dump.
```

Disasm after plugin load
```
Dump of assembler code for function _ZN13CInputManager22getMouseCoordsInternalEv:
   0x000059e3a2b3edc0 <+0>:     movabs $0x7d4206008491,%rax
   0x000059e3a2b3edca <+10>:    jmp    *%rax
   0x000059e3a2b3edcc <+12>:    pop    %rax
   0x000059e3a2b3edcd <+13>:    nop
   0x000059e3a2b3edce <+14>:    nop
   0x000059e3a2b3edcf <+15>:    nop
   0x000059e3a2b3edd0 <+16>:    nop
   0x000059e3a2b3edd1 <+17>:    nop
   0x000059e3a2b3edd2 <+18>:    nop
   0x000059e3a2b3edd3 <+19>:    nop
   0x000059e3a2b3edd4 <+20>:    nop
   0x000059e3a2b3edd5 <+21>:    mov    %rbx,0x8(%rsp)
   0x000059e3a2b3edda <+26>:    mov    %rdi,%rbx
   0x000059e3a2b3eddd <+29>:    call   0x59e3a2b02d30 <_ZN15CPointerManager8positionEv>
   0x000059e3a2b3ede2 <+34>:    mov    0x8(%rsp),%rax
   0x000059e3a2b3ede7 <+39>:    sub    %fs:0x28,%rax
   0x000059e3a2b3edf0 <+48>:    jne    0x59e3a2b3edfb <_ZN13CInputManager22getMouseCoordsInternalEv+59>
   0x000059e3a2b3edf2 <+50>:    add    $0x10,%rsp
   0x000059e3a2b3edf6 <+54>:    mov    %rbx,%rax
   0x000059e3a2b3edf9 <+57>:    pop    %rbx
   0x000059e3a2b3edfa <+58>:    ret
   0x000059e3a2b3edfb <+59>:    call   0x59e3a2887980 <__stack_chk_fail@plt>
End of assembler dump.
```

Disasm of trampoline (address from logs):
```
(gdb) x/16i 0x403bc100
   0x403bc100:  push   %rbx
   0x403bc101:  sub    $0x10,%rsp
   0x403bc105:  mov    0x62c3b6ac(%rip),%rsi        # 0xa2ff77b8
   0x403bc10c:  mov    %fs:0x28,%rbx
   0x403bc115:  push   %rax
   0x403bc116:  movabs $0x59e3a2b3edcc,%rax
   0x403bc120:  jmp    *%rax
   0x403bc122:  add    %al,(%rax)
   0x403bc124:  add    %al,(%rax)
   0x403bc126:  add    %al,(%rax)
   0x403bc128:  add    %al,(%rax)
   0x403bc12a:  add    %al,(%rax)
   0x403bc12c:  add    %al,(%rax)
   0x403bc12e:  add    %al,(%rax)
   0x403bc130:  add    %al,(%rax)
   0x403bc132:  add    %al,(%rax
```

**Comment by vaxerski** [comment]
hahaha okay so we got why it crashes :) thanks for the disasm, will try to fix the crash.

The "not called" is likely unfixable, unless we explicitly mark the fn as no inline. I am unsure if that's a good idea though.

**Comment by raybbian** [comment]
> The "not called" is likely unfixable, unless we explicitly mark the fn as no inline. I am unsure if that's a good idea though.

Unfortunate :( to the workarounds I go!

> hahaha okay so we got why it crashes :) thanks for the disasm, will try to fix the crash.

Would love a short explanation if you have the time :heart:

**Comment by vaxerski** [comment]
> Would love a short explanation if you have the time ❤️

You have some weird page allocations before the code sector so hyprland tries to use them as anchors for the trampolines, where it should use the code sections. Fixed that, should not crash anymore.

**Comment by raybbian** [comment]
I can confirm that calling the original function no longer crashes on non-nested release git. 

Interestingly, compiling Hyprland with (in `InputManager.cpp`)
```cpp
Vector2D __attribute__((noinline)) CInputManager::getMouseCoordsInternal() {
    return g_pPointerManager->position();
}
```
and (in `InputManager.hpp`)
```cpp
Vector2D __attribute__((noinline)) getMouseCoordsInternal();
```
still results in my function hook only executing on mouse click. Shouldn't this prevent an inline? The onMouseMove event hook is triggered and my callback is called, but my hook for `getMouseCoordsInternal` is not.

For documentation: I've gotten around this by using `g_pPointerManager->warpTo(...)` before and after calling interal fns like `onBeginWindowDrag`, which works for my use case. Curious as to why the hook doesn't get called still...

**Comment by vaxerski** [comment]
attach gdb and put a breakpoint on `_ZN13CInputManager22getMouseCoordsInternalEv`, check if it gets called... if not, I assume it got inlined.

**Comment by raybbian** [comment]
Doesn't break on mouse move, but breaks on mouse click - probably inlined.

Thanks for the help!
